### PR TITLE
twister: support test plans and test configuration

### PR DESCRIPTION
**Pull request by nashif**
Add support test plans and the ability to assign a specific test to one
or more plans. Use command line options of twister it is then possible
to select a plan and just execute the tests included in this plan.

Additionally, a test configuration is now available to define plan
dependencies and additional inclusion of tests into a specific plan if
the test itself does not have this information already.

In the configuration file you can include complete components using
regular expressions and you can specific which test plan to import from
the same file, making management of plans easier.

To help with testing outside of upstream CI infrastructure, additional
options are available in the configuration file, which can be hosted
locally. As of now, those options are available:

- Ability to ignore default platforms as defined in board definitions
  (Those are mostly emulation platforms used to run tests in upstream
  CI)
- Option to specify your own list of default platforms overriding what
  upstream defines.
- Ability to override build_onl_all options used in some testscases.
  This will treat tests or sample as any other just build for default
  platforms you specify in the configuation file or on the command line.

commits:
- twister: add support for plans and test configuration
- tests: add smoke tests
- tests: mem_map: remove scnearios used for coverage
- tests: mem_protect: cleanup yaml and reenable qemu_arc

### CODE REVIEW

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/testplan.py:184]: do we need check the uniq id here? I guess same cases maybe include multiply times

Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:184]: yes, I guess we can do the uniqueness check here.

**Comment by teburd** [comment]
@PerMac @hakehuang

**Review by hakehuang — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Comment by PerMac** [comment]
I am finalizing my review, please wait with merging.

**Comment by carlescufi** [comment]
DNM waiting for a review from @PerMac

**Review by PerMac — CHANGES_REQUESTED** [review]
In general I think this is a very useful feature. However, I have some objections.
Comments added. Additionally, since this is a new feature a proper documentation is IMO a must. I tested some workflows but often had to guess what is the intended behavior.

File: scripts/schemas/twister/testsuite-schema.yaml
Comment (nashif) [reply scripts/schemas/twister/testsuite-schema.yaml:57]: > Why such strict list of possible names?

The list can always be expanded. Otherwise we will end up with a mess with custom levels and things will be missed. This is a list of possible levels, does not mean you have to use them all, what fall under each of those is more or less well defined in the testing universe. 

This change has been there for 3 months and the list was open for discussion, but nobody complained. We can modify it expand on it, I am fine with that, but we need to have this pre-defined to avoid random levels going in that will be difficult to manage later.

pr, push, schedule are not testing levels, those are events and everyone can decide what to run and when based on their needs.

The confusion here is coming most likely from plan vs level, I will change everything back to level as it was before, but the schema will have to remain, maybe modifed a bit and following some standard names and what we need as a project.

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:435]: sure.

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:722]: yes, that was planned after initial agreement on the concept.

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:75]: This just corrosponds to what we have in the testcase schema, we can change this to make it more clear, np.

File: tests/test_config.yaml
Comment (nashif) [reply tests/test_config.yaml:3]: I guess you are misunderstanding the meaning of `default` platforms. It does not mean you only test on one single platform, it means those platforms are run with full scope, i.e. the same way we have `default: true` set in the platform yaml. If a test has filters that does not include any default platforms, it will run whatever appears in the plaform_allow. There is no change in behavior from the detauld twister configuration, here we remove default from what upstream defines and apply our own, the additional filtering is just the same as before.

If you only want to run `nrf52840dk_nrf52840`, use `-p nrf52840dk_nrf52840`

Comment (nashif) [reply tests/test_config.yaml]: the default behavior will be overriden using `increased_platform_scope: false`

Comment (PerMac) [reply tests/test_config.yaml]: My bad. Indeed, It works like default platforms. I didn't look close to realize that that these were platforms from platform_allow due to no match with default.

**Review by hakehuang — DISMISSED** [review]
I like the test level feature.

File: scripts/pylib/twister/twisterlib/environment.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/environment.py:275]:
```suggestion
        help="Test level to be used, by default, no levels are used for filtering"
```

**Review by PerMac — DISMISSED** [review]
just pointing 2 minor typos. I don't mind having them fixed latter on.

**Review by PerMac — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**
